### PR TITLE
Introduce SPI to swap authentication directives

### PR DESCRIPTION
**Pull request by mhenke1**
This PR introduces a SPI to configure a different authentication directive used in the REST API

## Description
This PR introduces a SPI to swap/exchange authentication directives in the REST API. 
The existing SPI does basic authentication only. With this PR alternative authentication methods 
like bearer token can be enabled. This capability is for example needed to support 
external IAM systems.

Additionally it packages the existing functionality to verify the basic authentication header to be consumed in the SPI without functional changes.

[PG 1:3115 passed]

## Related issue and scope
I explained the background of these changes on the mailing list
[Extending Authentication and Entitlement - Heads up](https://lists.apache.org/thread.html/891855d699710ba53e89a12c817c6e7d2560023c588f6f5a3bd74ca6@%3Cdev.openwhisk.apache.org%3E)

## My changes affect the following components
- [x] API
- [x] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [x] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:
- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

File: core/controller/src/main/scala/whisk/core/controller/BasicAuthenticationDirective.scala [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class AuthenticatedRouteBasicAuth(implicit val actorSystem: ActorSystem,
+                                  implicit val httpRequest: HttpRequest,
+                                  implicit val materializer: ActorMaterializer,
+                                  implicit val logging: Logging)
+    extends BasicAuthenticate {
+  protected implicit val executionContext = actorSystem.dispatcher
+  protected val authStore = WhiskAuthStore.datastore
```
Comment (chetanmeh) [inline core/controller/src/main/scala/whisk/core/controller/BasicAuthenticationDirective.scala]: `WhiskAuthStore.datastore` would use `SpiLoader` to load and create `ArtifactStore` for each request via reflection. As AuthStore does not change post startup it would be better to load it in `BasicAuthenticationDirectiveProvider` and pass it. 

May be make `BasicAuthenticate` an object and pass required dependencies as `validateCredentials` params.

**Comment by mhenke1** [comment]
@chetanmeh with the last commit ,the authStore is now created once and passed implicitly into the directive.

File: core/controller/src/main/scala/whisk/core/controller/BasicAuthenticationDirective.scala [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import whisk.common.{Logging, TransactionId}
+import whisk.core.entity._
+import whisk.core.entity.types.AuthStore
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class AuthenticatedRouteBasicAuth(implicit val authStore: AuthStore,
+                                  implicit val actorSystem: ActorSystem,
+                                  implicit val httpRequest: HttpRequest,
+                                  implicit val materializer: ActorMaterializer,
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/BasicAuthenticationDirective.scala]: As per f2f discussion, please have a look at various `extractX` methods to combine the directives and avoid broad interfaces.

File: core/controller/src/main/scala/whisk/core/controller/BasicAuthenticationDirective.scala [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+object BasicAuthenticationDirectiveProvider extends AuthenticationDirectiveProvider {
+
+  override def instance(implicit transid: TransactionId,
+                        authStore: AuthStore,
+                        httpRequest: HttpRequest,
+                        actorSystem: ActorSystem,
+                        materializer: ActorMaterializer,
+                        logging: Logging): AuthenticationDirective[Identity] =
+    new AuthenticatedRouteBasicAuth().getDirective
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/BasicAuthenticationDirective.scala]: I think, given all the `extractX` directives, we can model this to be a static directive which only needs to be instantiated once.

File: core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala [hunk 4]
```diff
@@ -191,36 +194,38 @@ class RestAPIVersion(config: WhiskConfig, apiPath: String, apiVersion: String)(
         "swagger_paths" -> JsObject("ui" -> s"/$swaggeruipath".toJson, "api-docs" -> s"/$swaggerdocpath".toJson)))
   }
 
-  def routes(implicit transid: TransactionId): Route = {
+  def routes(implicit transid: TransactionId, authStore: AuthStore): Route = {
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala]: Do you need the authStore here? It's already in scope I think?

File: core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala [hunk 5]
```diff
@@ -191,36 +194,38 @@ class RestAPIVersion(config: WhiskConfig, apiPath: String, apiVersion: String)(
...
-            }
-        } ~ {
-          swaggerRoutes
-        }
-      } ~ {
-        // web actions are distinct to separate the cors header
-        // and allow the actions themselves to respond to options
-        basicAuth(validateCredentials) { user =>
-          web.routes(user)
+      extractRequest { httpRequest =>
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala]: Why is this needed?

**Comment by mhenke1** [comment]
[PG 3:2532 is ✅ ]
